core/jobs/batch_jobs/audit_rule_validation_jobs.py

```diff
@@ -0,0 +1,2194 @@
+        We will perform this by checking the number of invalid answer
+        groups and the number of answer groups having destination other than
+        try again are equal then we will store the state name and ans group
+
+        Args:
+            states_dict: dict[str, state_domain.State]. The state dictionary.
+
+        Returns:
+            invalid_states: List[Dict[str, object]]. List of invalid states
+            with the invalid answer groups.
+        """
+        invalid_states = []
+        for state_name, state in states_dict.items():
+            if state.interaction.id != 'NumericInput':
+                continue
+            answer_groups = state.interaction.answer_groups
+            invalid_ans_group_count = 0
+            valid_dest_ans_group_count = 0
+            invalid_ans_groups = []
+            rule_spec_till_now: List[state_domain.RuleSpecDict] = []
+            ranges: List[RangeVariable] = []
+            for ans_group_idx, answer_group in enumerate(answer_groups):
+                can_delete_ans_group, ranges = (
+                    ExpAuditRuleChecksJob.
+                    _filter_invalid_numeric_ans_group(
+                        answer_group, rule_spec_till_now, ans_group_idx,
+                        ranges
+                    )
+                )
+
+                for rule_spec in answer_group.rule_specs:
+                    rule_spec_till_now.append(rule_spec.to_dict())
+
+                if answer_group.outcome.dest != state_name:
+                    valid_dest_ans_group_count += 1
+                    if can_delete_ans_group:
+                        invalid_ans_group_count += 1
+                        invalid_ans_groups.append(ans_group_idx)
+
+            if (
+                invalid_ans_group_count == valid_dest_ans_group_count and
+                invalid_ans_group_count != 0 and
+                valid_dest_ans_group_count != 0
+            ):
+                invalid_states.append(
+                    {
+                        'state_name': state_name,
+                        'ans_group_idx': invalid_ans_groups
+                    }
+                )
+
+        return invalid_states
+
+    @staticmethod
+    def _should_check_range_criteria(
+        earlier_rule: state_domain.RuleSpec,
+        later_rule: state_domain.RuleSpec
+    ) -> bool:
+        """Checks the range criteria between two rules by comparing their
+        rule type
+
+        Args:
+            earlier_rule: state_domain.RuleSpec. Previous rule.
+            later_rule: state_domain.RuleSpec. Current rule.
+
+        Returns:
+            bool. Returns True if the rules passes the range criteria check.
+        """
+        if earlier_rule.rule_type != 'IsExactlyEqualTo':
+            return True
+        return later_rule.rule_type not in (
+            'IsExactlyEqualTo', 'IsEquivalentTo',
+            'IsEquivalentToAndInSimplestForm'
+        )
```

Comment:
Make sure your conversion function doesn't use `not in`. If a new rule type is added, then it may fail. Instead, you must use `in` and put the rules that you want to be checked.

Reply:
Acknowledged!